vk/backend/backend.go

```diff
@@ -113,169 +116,302 @@ func state2containerState(prevState *v1.ContainerState, currState titusdriver.Ti
 	}
 }
 
-func RunWithBackend(ctx context.Context, rp runtimeTypes.ContainerRuntimeProvider, m metrics.Reporter, statuses *os.File, pod *v1.Pod, cfg config.Config) error { // nolint: gocyclo
+type Backend struct {
+	network, disk, memory, gpu, cpu resource.Quantity
+	pod                             *v1.Pod
+	containerinfo                   *titus.ContainerInfo
+	readyErr                        error
+	readyLock                       sync.RWMutex
+	m                               metrics.Reporter
+	rp                              runtimeTypes.ContainerRuntimeProvider
+	cfg                             *config.Config
+}
+
+func getContainerInfo(pod *v1.Pod) (string, error) {
 	containerInfoStr, ok := pod.GetAnnotations()["containerInfo"]
-	if !ok {
-		return errContainerInfo
+	if ok {
+		return containerInfoStr, nil
 	}
-	data, err := base64.StdEncoding.DecodeString(containerInfoStr)
-	if err != nil {
-		return errors.Wrap(err, "Could not decode containerInfo from base64")
-	}
-	var containerInfo titus.ContainerInfo
 
-	err = proto.Unmarshal(data, &containerInfo)
-	if err != nil {
-		return errors.Wrap(err, "Could not deserialize protobuf")
+	containerInfoStr, ok = pod.GetAnnotations()[podCommon.AnnotationKeyPodTitusContainerInfo]
+	if !ok {
+		return "", errContainerInfo
 	}
 
-	useBytes, err := podCommon.ByteUnitsEnabled(pod)
+	return containerInfoStr, nil
+}
+
+func NewBackend(ctx context.Context, rp runtimeTypes.ContainerRuntimeProvider, pod *v1.Pod, cfg *config.Config, m metrics.Reporter) (*Backend, error) {
+	var containerInfo titus.ContainerInfo
+
+	podSchemaVer, err := podCommon.PodSchemaVersion(pod)
 	if err != nil {
-		return errors.Wrap(err, "Could not parse byte units pod annotation")
+		return nil, err
 	}
 
-	limits := pod.Spec.Containers[0].Resources.Limits
+	// As of pod schema v1, the containerInfo annotation is optional.
+	if podSchemaVer < 1 {
+		containerInfoStr, err := getContainerInfo(pod)
+		if err != nil {
+			return nil, err
+		}
 
-	// TODO: pick one, agreed upon resource name after migration to k8s scheduler is complete.
-	disk, _ := resource.ParseQuantity("2G")
-	for _, k := range []v1.ResourceName{v1.ResourceEphemeralStorage, v1.ResourceStorage, "titus/disk"} {
-		if v, ok := limits[k]; ok {
-			disk = v
-			break
+		data, err := base64.StdEncoding.DecodeString(containerInfoStr)
+		if err != nil {
+			return nil, errors.Wrap(err, "Could not decode containerInfo from base64")
```

Comment:
Use fmt.Errorf now.

Reply:
I don't want to drift from what we have "upstream" in master.

Keep in mind I didn't write this code, I just copy/pasted this file from what we already had.